upstairs/src/dummy_downstairs_tests.rs

```diff
@@ -1880,4 +1881,710 @@ pub(crate) mod protocol_test {
 
         Ok(())
     }
+
+    /// Test that an error during the live repair doesn't halt indefinitely
+    #[tokio::test]
+    async fn test_error_during_live_repair_no_halt() -> Result<()> {
+        let harness = Arc::new(TestHarness::new().await?);
+
+        let (jh1, mut ds1_messages) =
+            harness.ds1().await.spawn_message_receiver().await;
+        let (_jh2, mut ds2_messages) =
+            harness.ds2.spawn_message_receiver().await;
+        let (_jh3, mut ds3_messages) =
+            harness.ds3.spawn_message_receiver().await;
+
+        // Send 200 more than IO_OUTSTANDING_MAX jobs. Flow control will kick in
+        // at MAX_ACTIVE_COUNT messages, so we need to be sending read responses
+        // while reads are being sent. After IO_OUTSTANDING_MAX jobs, the
+        // Upstairs will set ds1 to faulted, and send it no more work.
+        const NUM_JOBS: usize = IO_OUTSTANDING_MAX + 200;
+        let mut job_ids = Vec::with_capacity(NUM_JOBS);
+
+        for i in 0..NUM_JOBS {
+            info!(harness.log, "sending read {}/{NUM_JOBS}", i);
+
+            {
+                let harness = harness.clone();
+
+                // We must tokio::spawn here because `read` will wait for the
+                // response to come back before returning
+                tokio::spawn(async move {
+                    let buffer = Buffer::new(512);
+                    harness
+                        .guest
+                        .read(Block::new_512(0), buffer)
+                        .await
+                        .unwrap();
+                });
+            }
+
+            if i < MAX_ACTIVE_COUNT {
+                // Before flow control kicks in, assert we're seeing the read
+                // requests
+                bail_assert!(matches!(
+                    ds1_messages.recv().await.unwrap(),
+                    Message::ReadRequest { .. },
+                ));
+            } else {
+                // After flow control kicks in, we shouldn't see any more
+                // messages
+                match ds1_messages.try_recv() {
+                    Err(TryRecvError::Empty) => {}
+                    Err(TryRecvError::Disconnected) => {}
+                    x => {
```

Comment:
Do we need to ignore ping messages here too?

For all these cases really, for ds2 and ds3, they could all receive a ping, right?

Reply:
See other comment :)